### DISCUSSION TITLE
Debugging experience

### DISCUSSION THREAD

**Discussion by thoughtsunificator** [denoland/deno#27149] [General]
Start the debugging client as soon as a debugger statement is met. 

Information about the debugging protocols should be abstracted, which for the most part is not the case since we need to know that there is a server and a client, why should the developer care? How should they be abstracted? Well we can start with the fact we currently need to ask twice for the debugger to start, one with the debugger statement and two by using the debug command and in some cases you even have to add the inspect command to the equation, that's a hell lot more than just running your program and expecting the runtime to do what you asked by using the debugger statement which is... Start the debugger. 

**I propose that either a new env is added or we use `NODE_ENV` which seems like a pretty good use of it. When set, ANY script shall be debuggable either using the internal debugger client (CLI) or simply by providing the server depending on the execution context.**

If that's the case, then:

- `deno run foo.js`
- `deno task foo`
- `deno test`

All of these will work as is and be easily debuggable with no additional effort required (other than using the debugger statement) on the developer part.

**Comment by marvinhagemeister** [comment]
FYI: Both Deno and Node are built around V8 JS engine which is why the debugger integration is the same. It's provided by V8, not by Deno or Node. So there is no copying anywhere because both projects directly use the one provided by V8.

**Reply by thoughtsunificator** [reply]
V8 gives you the [inspector protocol](https://v8.dev/docs/inspector), which you can build a client on. Deno "built" the exact same client as https://github.com/node-inspector/node-inspector did, and in the exact same wrong way. And the funniest part being that [Bun](https://github.com/oven-sh/bun) also [did the exact same thing](https://bun.sh/docs/runtime/debugger#inspect)  as Deno did...

**Reply by thoughtsunificator** [reply]
I'm sorry, am I missing something here? Are debuggers a taboo? I've never heard of a religion against debugging though... 

I guess that's it for deno.

**Reply by bartlomieju** [reply]
> Start the debugger (CLI) as soon as the debugger keyword is met, it can't be that impossible that even 3 runtimes later we can't have a reasonable debugging experience.

The way V8 debugger works is you need to have client connected to stop on a `debugger` keyword, and it's already possible when you're running code from your IDE like VSCode or JetBrains. You can start a debugging session and when the `debugger` keyword is hit they will pause execution and let you inspect the state of program.

**Reply by thoughtsunificator** [reply]
I believe I understand some parts of the architecture of it which lead me to make it clear that these are concerns that should be addressed on the runtime side and certainly not IDE side or engine side for that particular matter.

I've read some parts of the source code of the current Node.js and I pretty much get why this sucks, although it'd be a little bit harder for me to offer a solution to this (vital) problem plus the fact that I don't want to ask anyone to refactor its entire codebase for such a core feature that would only, be... mean.

So basically, any developer that expects any form of debugging experience through the most primitive tools such as CLI is asked to develop (and maintain) its own debugger client and work on the interoperability of at least a dozen of tools across the toolchains, am I right?

**Reply by bartlomieju** [reply]
> So basically, any developer that expects any form of debugging experience through the most primitive tools such as CLI is asked to develop (and maintain) its own debugger client and work on the interoperability of at least a dozen of tools across the toolchains, am I right?

I believe Node has a CLI based debugger with `node inspect script.js`, Deno doesn't have it, because no one has ever asked for a CLI debugger. I'm not sure which dozen tools across toolchains you mean. I believe all JS runtimes understand and speak V8 debugging protocol, so that would be the only thing needed to support then.

**Reply by thoughtsunificator** [reply]
In the less bad scenario, you still need to explicitly ask the run time to activate its debugging client. which is exactly what I advocated against in the first place. 

This does not work well because you actually need to ask twice for the debugger client, one using the CLI and two by putting a debugger statement in your code, the other reason why this might not work so well is these other tools I talked about, they might not care about starting a debugger which mean anything that you build on top off breaks this precise behavior. 

At this point, Deno just brought the same failures from node-inspector.

**Reply by marvinhagemeister** [reply]
It seems like you have a well defined vision in your head of how you'd like the debugger to work. Have you considered making a PR?

**Reply by thoughtsunificator** [reply]
Well... Maybe I would if anyone could try and explain to me why it was done that way :)

**Reply by bartlomieju** [reply]
> Well, maybe I would if anyone could explain to me why it was done that way.

Because connecting a debugger client requires spawning a WebSocket server, which is not free to create and binds to a certain port (`127.0.0.1:9229` by default). Most users don't want a WS server being spawned in their production environment and that's why it's not done unless you provide a flag to inform you'd like to connect a debugger.

**Reply by thoughtsunificator** [reply]
Then how about an env instead? Which would, at the very least, fix my second issue (and the issue you mentioned). 
How about `NODE_ENV` for example, which already exists by the way. 

The bind thing is a non-issue, as Deno is already aware that both the client and the server originates from a single command: `deno run foo.js` in a `NODE_ENV` that is for sure not set to production.

I mean, that's the whole point of it, if we're trying to accommodate our DX to production needs we're not going to go far.

**I propose that either a new env is added or we use `NODE_ENV` which seems like a pretty good use of it. When set, ANY script shall be debuggable either using the internal debugger client (CLI) or simply by providing the server, depending on the execution context.**

**Reply by thoughtsunificator** [reply]
I edited both the title and the content to something that is a bit more on point.
